django_emarsys/util.py

```diff
@@ -0,0 +1,97 @@
+
+    if not getattr(settings, 'EMARSYS_BASE_URI', None):
+        log.warning("no EMARSYS_BASE_URI setting set")
+
+    events = getattr(settings, 'EMARSYS_EVENTS', None)
+    if not events:
+        log.warning("no EMARSYS_EVENTS setting set")
+        return
+
+    VALID_ARGUMENT = re.compile(r'^[a-z][a-z_0-9]*$')
+    for event, data in events.items():
+        if not event:
+            log.warning("invalid event name: '{event}'".format(event=event))
+
+        names = set()
+        args = set()
+        for data in data:
+            if 'name' not in data:
+                log.error("missing data key for event '{event}': 'name'"
+                          .format(event=event))
+
+            else:
+                name = data['name']
+                if name in names:
+                    log.warning(
+                        "reused data name for event '{event}': '{name}'"
+                        .format(name=name, event=event))
+
+                if not name or not name.strip():
+                    log.warning(
+                        "invalid data name for event '{event}': '{name}'"
+                        .format(name=name, event=event))
+
+                names.add(name)
+
+            if 'arg' not in data:
+                log.error("missing data key for event '{event}': 'arg'"
+                          .format(event=event))
+            else:
+                arg = data['arg']
+                if arg in args:
+                    log.error(
+                        "reused data argument for event '{event}': '{arg}'"
+                        .format(arg=arg, event=event))
+
+                if not arg or not VALID_ARGUMENT.match(arg):
+                    log.error(
+                        "invalid data argument for event '{event}': '{arg}'"
+                        .format(arg=arg, event=event))
+
+                args.add(arg)
+
+            if 'model' not in data:
+                log.error("missing data key for event '{event}': 'model'"
+                          .format(event=event))
+            else:
+                model = data['model']
+                chunks = model.split('.')
+                if len(chunks) < 2:
+                    log.error("model for event '{event}' expected to be "
+                              "'<app_label>.<model_name>', got: '{model}'"
+                              .format(event=event, model=model))
+                else:
+                    try:
+                        get_model(model)
+                    except LookupError as e:
+                        log.error("bad model '{model}' for event "
+                                  "'{event}': {error}"
+                                  .format(model=model, event=event, error=e))
+
+            for key in data:
+                if key not in ('name', 'arg', 'model'):
+                    log.warning(
+                        "unexpected data key for event '{event}': '{key}'"
+                        .format(event=event, key=key))
```

Comment:
I think that'd shorten some of the code, but result in less specific error/warning messages.
`{'name', 'arg', 'model'} != set(data)` would have to look into it in detail again or just output "there's a mismatch: <expected>, <actual>", which is less helpful.

Also, an additional keyword is no problem, so it's only a warning, but one missing is an error. 

Similarly, finding which `arg` or `name` is duplicated with the list you gave seems more complicated in the end, while I agree that the check "is anything duplicated?" is shorter then.

So if the same level of reporting detail is wished for (I wish for it!), then the code probably would turn out complicated again.

But I think there are some lines where `set` arithmetic will make the code less bulky and I'll try to reduce the boiler plate size.